File: app/src/main/java/io/caster/simplemvp/AppModule.java
package io.caster.simplemvp;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import io.caster.simplemvp.presentation.UserPresenter;
import io.caster.simplemvp.presentation.UserPresenterImpl;
import io.caster.simplemvp.repository.InMemoryUserRepositoryImpl;
import io.caster.simplemvp.repository.UserRepository;

@Module
public class AppModule {
    @Provides @Singleton
    public UserRepository provideUserRepository() {
        return new InMemoryUserRepositoryImpl();
    }

    @Provides
    public UserPresenter provideUserPresenter(UserRepository userRepository) {
        return new UserPresenterImpl(userRepository);
    }
}

File: app/src/main/java/io/caster/simplemvp/AppModule.kt
package io.caster.simplemvp

import dagger.Module
import dagger.Provides
import io.caster.simplemvp.presentation.UserPresenter
import io.caster.simplemvp.presentation.UserPresenterImpl
import io.caster.simplemvp.repository.InMemoryUserRepositoryImpl
import io.caster.simplemvp.repository.UserRepository
import javax.inject.Singleton

@Module
class AppModule {
    @Provides
    @Singleton
    fun provideUserRepository(): UserRepository {
        return InMemoryUserRepositoryImpl()
    }

    @Provides
    fun provideUserPresenter(userRepository: UserRepository): UserPresenter {
        return UserPresenterImpl(userRepository)
    }
}